### PR TITLE
Lecture 1 setup and product dao

### PR DESCRIPTION
**Pull request by trowqe**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,75 @@
...
 public class ArrayListProductDao implements ProductDao {
+
+    private List<Product> threadSaveArrayList;
+
+    public ArrayListProductDao() {
+        this.threadSaveArrayList = new CopyOnWriteArrayList<>();
+        initArrayList(threadSaveArrayList);
+    }
+
+    private List initArrayList(List list) {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: don't use raw types for a generic interface

Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: method should be `void`

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -1,25 +1,75 @@
...
+        list.add(new Product(9L, "nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
+        list.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
+        list.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
+        list.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
+        list.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+        return list;
+    }
+
+    public List<Product> getThreadSaveArrayList() {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: where is it used?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -1,25 +1,75 @@
...
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        Optional<Product> foundProduct =
+                threadSaveArrayList.stream()
+                        .filter(product -> product.getId() == id)
+                        .findFirst();
+        return foundProduct.get();
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: what would be if no product found?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 4]
```diff
@@ -1,25 +1,75 @@
...
+    public List<Product> getThreadSaveArrayList() {
+        return threadSaveArrayList;
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        Optional<Product> foundProduct =
+                threadSaveArrayList.stream()
+                        .filter(product -> product.getId() == id)
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: what's wrong is here?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -1,25 +1,75 @@
...
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        threadSaveArrayList.add(product);
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+        threadSaveArrayList.removeIf(product -> product.getId() == id);
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: what's wrong is here?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -1,25 +1,75 @@
...
-        throw new RuntimeException("Not implemented");
+        Optional<Product> foundProduct =
+                threadSaveArrayList.stream()
+                        .filter(product -> product.getId() == id)
+                        .findFirst();
+        return foundProduct.get();
     }
 
     @Override
+    //The method returns products having non null price and stock level > 0
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: use java doc

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java [hunk 7]
```diff
@@ -5,6 +5,7 @@
 public interface ProductDao {
     Product getProduct(Long id);
     List<Product> findProducts();
+    List<Product> userSearch(String criteria);
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: bad architectural decision

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 8]
```diff
@@ -13,29 +15,23 @@
 import java.util.List;
 
 public class ProductListPageServlet extends HttpServlet {
+    private static ProductDao productDao;
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: why is it static?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 9]
```diff
@@ -1,25 +1,75 @@
...
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        return threadSaveArrayList.stream()
+                .filter(product -> product.getStock() > 0)
+                .filter(product -> product.getPrice() != null)
+                .collect(Collectors.toList());
+    }
+
+    public List<Product> userSearch(String userInput){
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: remove from the current pull request

File: src/main/java/com/es/phoneshop/model/product/ProductDao.java (hunk 7 above)
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/model/product/ProductDao.java]: remove from the current pull request

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 10]
```diff
@@ -13,29 +15,23 @@
...
+
+    @Override
+    public void init() throws ServletException {
+        productDao = new ArrayListProductDao();
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: what bad is here?

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 11]
```diff
@@ -13,29 +15,23 @@
...
-        result.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
-        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
-
-        return result;
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
Comment (CaH9I) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java]: remove from the current pull request

File: src/main/webapp/WEB-INF/pages/productList.jsp [hunk 12]
```diff
@@ -8,6 +8,12 @@
   <p>
     Welcome to Expert-Soft training!
   </p>
+
+  <form method="post">
```
Comment (CaH9I) [inline src/main/webapp/WEB-INF/pages/productList.jsp:12]: remove from the current pull request

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 13]
```diff
@@ -15,7 +17,40 @@ public void setup() {
     }
 
     @Test
-    public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+    public void deleteProductTest() {
+        Product testProduct = new Product(23L, "simsxg75", "Siemens SXG75", new BigDecimal(150), Currency.getInstance("USD"), 40,
+                "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg");
+        productDao.delete(23L);
+        assertEquals(productDao.getProduct(23L), testProduct);
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: what does this `assertEquals` check?

File: src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java [hunk 14]
```diff
@@ -27,15 +27,16 @@
...
+    public void whenCallDoGetTheServletReturnProductListJsp () throws ServletException, IOException {
+         String path = "/WEB-INF/pages/productList.jsp";
+         when(request.getRequestDispatcher(path)).thenReturn(requestDispatcher);
+
+         servlet.init();
+         servlet.doGet(request, response);
 
-        verify(requestDispatcher).forward(request, response);
+         verify(request, times(1)).getRequestDispatcher(path);
+         verify(request,never()).getSession();
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java]: why do you verify `getSession()`?

File: src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java [hunk 15]
```diff
@@ -27,15 +27,16 @@
...
-        servlet.doGet(request, response);
+    public void whenCallDoGetTheServletReturnProductListJsp () throws ServletException, IOException {
+         String path = "/WEB-INF/pages/productList.jsp";
+         when(request.getRequestDispatcher(path)).thenReturn(requestDispatcher);
+
+         servlet.init();
+         servlet.doGet(request, response);
 
-        verify(requestDispatcher).forward(request, response);
+         verify(request, times(1)).getRequestDispatcher(path);
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java]: why do you verify `getRequestDispatcher()`?

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 16]
```diff
@@ -15,7 +17,40 @@ public void setup() {
...
+    public void saveProductTest() {
+        Product testProduct = new Product(43L, "simsxg75", "Siemens SXG75", new BigDecimal(150), Currency.getInstance("USD"), 40,
+                "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg");
+        productDao.save(testProduct);
+        assertEquals(productDao.getProduct(43L), testProduct);
+    }
+
+
+    @Test
+    public void findProductsTest() {
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: create two separate test methods

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 17]
```diff
@@ -15,7 +17,40 @@ public void setup() {
...
+        Product testProductWithNullPrice = new Product(53L, "simsxg75", "Siemens SXG75", null, Currency.getInstance("USD"), 40,
+                "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg");
+        productDao.save(testProductWithNullPrice);
+        assertFalse(productDao.findProducts().contains(testProductWithNullPrice));
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:59]: what's wrong?

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 18]
```diff
@@ -15,7 +17,40 @@ public void setup() {
...
+
+        Product testProductWithStockLevel0 = new Product(63L, "simsxg75", "Siemens SXG75", null, Currency.getInstance("USD"), 0,
+                "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg");
+        productDao.save(testProductWithNullPrice);
+        assertFalse(productDao.findProducts().contains(testProductWithStockLevel0));
```
Comment (CaH9I) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:67]: what's wrong?

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java (hunk 17 above)
Comment (trowqe) [reply src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:59]: BigDecimal nullPrise = null;
        Product testProductWithNullPrice = new Product(53L, "simsxg75", "Siemens SXG75", nullPrise, Currency.getInstance("USD"), 40,
                "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg");

Comment (CaH9I) [reply src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:59]: Hint: in real application in your case the assert statement would always be `false`. That means your test would never fail regardless the products you have in the storage. Why?

**Pull request merged**